File: CS207b/L/w01/improved/model/Movie.py
class PriceCode:
    __code = None
    __days_before_additional_charge = None
    __additional_charge_rate = None

    def __init__(self, code, days_before_additional_charge, additional_charge_rate):
        self.__code = code
        self.__days_before_additional_charge = days_before_additional_charge
        self.__additional_charge_rate = additional_charge_rate

    def get_code(self):
        return self.__code

    def set_code(self, arg):
        self.__code = arg

    def get_days_before_additional_charge(self):
        return self.__days_before_additional_charge

    def get_additional_charge_rate(self):
        return self.__additional_charge_rate

    def get_charge(self, days_rented):
        if days_rented <= self.__days_before_additional_charge:
            return days_rented
        else:
            return (
                self.__days_before_additional_charge
                + (days_rented - self.__days_before_additional_charge)
                * self.__additional_charge_rate
            )


class Movie:
    __title = None
    __price_code = None

    REGULAR = PriceCode(0, 2, 1.5)
    NEW_RELEASE = PriceCode(1, 0, 3)
    CHILDRENS = PriceCode(2, 3, 1.5)

    def __init__(self, title, priceCode):
        self.__title = title
        self.__price_code = priceCode

    def get_price_code(self):
        return self.__price_code.get_code()

    def set_price_code(self, priceCode):
        self.__price_code = priceCode

    def get_title(self):
        return self.__title

    def get_charge(self, days_rented):
        return self.__price_code.get_charge(days_rented)

File: CS207b/L/w04/pkg/movie/movie.go
package movie

type MovieType struct {
	index                    int
	initialCharge            float64
	baseDailyRate            float64
	daysBeforeAdditionCharge float64
	additionalDailyRate      float64
}

func NewMovieType(index int, initialCharge, baseDailyRate, daysBeforeAdditionCharge, additionalDailyRate float64) MovieType {
	return MovieType{index: index, initialCharge: initialCharge, baseDailyRate: baseDailyRate, daysBeforeAdditionCharge: daysBeforeAdditionCharge, additionalDailyRate: additionalDailyRate}
}

type Movie struct {
	title     string
	movieType MovieType
}

func NewMovie(title string, movieType MovieType) Movie {
	return Movie{title: title, movieType: movieType}
}

func (m *Movie) GetTitle() string {
	return m.title
}

func (m *Movie) SetMovieType(movieType MovieType) {
	m.movieType = movieType
}

func (m *Movie) GetPriceCode() int {
	return m.movieType.index
}

func (m *Movie) SetPriceCode(arg int) {
	m.movieType.index = arg
}

func (m *Movie) GetCharge(daysRented int) float64 {
	daysRentedFloat := float64(daysRented)
	result := m.movieType.initialCharge
	if daysRentedFloat > m.movieType.daysBeforeAdditionCharge {
		result += (daysRentedFloat - m.movieType.daysBeforeAdditionCharge) * m.movieType.additionalDailyRate
	} else {
		result += m.movieType.baseDailyRate * daysRentedFloat
	}
	return result
}
